include/libopencm3/stm32/f4/dma2d.h

```diff
@@ -0,0 +1,222 @@
+/*
+ * STM32F4 DMA2D Register defines
+ *
+ * Copyright (C) 2016, Chuck McManis <cmcmanis@mcmanis.com>
+ *
+ * This file is part of the libopencm3 project.
+ *
+ * This library is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include <libopencm3/stm32/memorymap.h>
+#ifndef uint32_t
+#include <stdint.h>
+#endif
+
+
+#ifndef DMA2D_H
+#define DMA2D_H
+/*
+ * A couple of helper macros to avoid copy pasta code
+ * For multi-bit fields we can use these macros to create
+ * a unified SET_<reg>_<field> version and a GET_<reg_<field> version.
+ */
+#define DMA2D_GET(reg, field, x) \
+    (((x) >> DMA2D_ ## reg ## _ ## field ##_SHIFT) & DMA2D_ ## reg ##_ ## field ##_MASK)
+
+#define DMA2D_SET(reg, field, x) \
+    (((x) & DMA2D_ ## reg ##_## field ##_MASK) << DMA2D_## reg ##_## field ##_SHIFT)
+
+/* DMA2D Control Register */
+#define DMA2D_CR			MMIO32(DMA2D_BASE + 0x0U)
+#define DMA2D_CR_MODE_SHIFT		16
+#define DMA2D_CR_MODE_MASK		0x3
+#define DMA2D_CR_MODE_M2M		0	/* memory to memory */
+#define DMA2D_CR_MODE_M2MWPFC		1	/* memory to memory with pix convert */
+#define DMA2D_CR_MODE_M2MWB		2	/* memory to memory with blend */
+#define DMA2D_CR_MODE_R2M		3	/* register to memory */
+#define DMA2D_CR_CEIE			(1 << 13)
+#define DMA2D_CR_CTCIE			(1 << 12)
+#define DMA2D_CR_CAEIE			(1 << 11)
+#define DMA2D_CR_TWIE			(1 << 10)
+#define DMA2D_CR_TCIE			(1 << 9)
+#define DMA2D_CR_TEIE			(1 << 8)
+#define DMA2D_CR_ABORT			(1 << 2)
+#define DMA2D_CR_SUSP			(1 << 1)
+#define DMA2D_CR_START			(1 << 0)
+
+/* DMA2D Interrupt Status Register */
+#define DMA2D_ISR			MMIO32(DMA2D_BASE + 0x4U)
+#define DMA2D_ISR_CEIF			(1 << 5)
+#define DMA2D_ISR_CTCIF			(1 << 4)
+#define DMA2D_ISR_CAEIF			(1 << 3)
+#define DMA2D_ISR_TWIF			(1 << 2)
+#define DMA2D_ISR_TCIF			(1 << 1)
+#define DMA2D_ISR_TEIF			(1 << 0)
+
+/* DMA2D Interrupt Flag Clear Register */
+#define DMA2D_IFCR			MMIO32(DMA2D_BASE + 0x8U)
+#define DMA2D_IFCR_CCEIF		(1 << 5)
+#define DMA2D_IFCR_CCTCIF		(1 << 4)
+/* this is what the reference manual specifies */
+#define DMA2D_IFCR_CAECIF		(1 << 3)
+/* this is what it "should be" */
+#define DMA2D_IFCR_CCAEIF		(1 << 3)
```

Comment:
Actually I know for a fact that several will be fixed, as I was doing my code I was feeding documentation bugs to one of ST's field engineers who assured me they would get to the right technical pubs people.